test: Get rid of 4.9 pipeline

brb opened the pull request:
In v1.13 we stopped supporting the 4.9 kernel \[1\]. Therefore, we can remove the 4.9 CI pipeline for the main and v1.13 branches.

Previously, the 4.9 CI used to run with kube-proxy and KPR=disabled, while the 4.19 CI with kube-proxy and KPR=strict (aka mixed mode). This commit sets KPR=disabled in the 4.19 CI. The mixed mode is tested in the new ci-datapath GHA (minus the IPv6 until \[2\] has been resolved; this is fine).

\[1\]: https://github.com/cilium/cilium/issues/22116
\[2\]: https://github.com/cilium/cilium-cli/issues/1139

joestringer commented on test/helpers/kubectl.go:
```diff
@@ -2484,14 +2484,14 @@ func (kub *Kubectl) overwriteHelmOptions(options map[string]string) error {
 		addIfNotOverwritten(options, "sessionAffinity", "true")
 	}
 
 	// Disable unsupported features that will just generated unnecessary
 	// warnings otherwise.
```
Seems like now the decision isn't about available support, but actually about how we're choosing to configure Cilium depending on the environment:

```suggestion
```

joestringer replied:
Also, the change makes this condition and the next conditional check exactly the same so we can combine the two. Same for the conditional check immediately above.

brb replied:
> how we're choosing to configure Cilium depending on the environment

I think that was always the case for the e2e ginkgo tests. The configuration was implicitly based on mostly `KERNEL` and `KUBEPROXY` env vars.

joestringer replied:
I messed up the code suggestion upon first try above, I was just pointing out general code cleanups to make this easier to read. I see what you mean about the right path forward being just to drop the code entirely instead of spending more time iterating on nits like this. Makes sense :+1:

joestringer commented on test/k8s/hubble.go:
```diff
@@ -25,9 +25,9 @@ import (
 var _ = Describe("K8sAgentHubbleTest", func() {
 	// We want to run Hubble tests both with and without our kube-proxy
 	// replacement, as the trace events depend on it. We thus run the tests
-	// on GKE and our 4.9 pipeline.
+	// on GKE and our 4.19 pipeline.
 	SkipContextIf(func() bool {
-		return helpers.RunsOn419OrLaterKernel() || helpers.RunsOnAKS()
+		return helpers.RunsOnNetNextKernel() || helpers.RunsOnAKS()
```
Shouldn't this use `RunsWithKubeProxyReplacement()` to satisfy the comment? Then we don't need to keep updating which kernels to run here or the comment.

joestringer replied:
Actually now that 4.19 is the minimum, can we reuse `RunsWithKubeProxyReplacement()` more widely for other cases that are updated in this PR? It would be nice if we didn't have so many kernel version dependencies scattered throughout the test code.

brb replied:
I thought about doing the `helpers.Runs*()` refactoring so that the skip conditions / configuration choices were more explicit and consistent.

However, I've started considering `test/k8s/*` to be a dead code which should be touched as little as possible, and which will be removed immediately after we have finished extending the coverage of `ci-datapath`.

pchaigno replied:
The first time someone told me to stop investing time into the ginkgo suite was 2 years ago (shortly after the CLI came out). I'm glad I didn't listen.

joestringer commented on test/k8s/service_helpers.go:
```diff
@@ -865,13 +867,17 @@ func testExternalTrafficPolicyLocal(kubectl *helpers.Kubectl, ni *helpers.NodesI
-		testCurlFromPodInHostNetNS(kubectl, httpURL, count, 0, ni.K8s1NodeName)
-		testCurlFromPodInHostNetNS(kubectl, tftpURL, count, 0, ni.K8s1NodeName)
+
+		// Until https://github.com/cilium/cilium/issues/23481 has been fixed
+		if !ipsec {
```
This should ideally be in a separate commit so we can (a) store the rationale and (b) just revert it later when we fix #23481.

joestringer replied:
It would also help to document why specific callers of this function pass true or false. At a glance I couldn't quite establish why some sibling statements below pass ipsec=true or ipsec=false.

brb replied:
My reasoning https://github.com/cilium/cilium/pull/23343#discussion_r1095662225 applies here too :-)

I hope we will add the externalTrafficPolicy=Local test cases to the `ci-datapath` sooner than #23481 gets fixed.

Do you still want me to move this to a separate commit?

joestringer replied:
Not splitting the changes into separate commits mainly has the impacts of making the reviewer job harder, which is why I always encourage more effort in that regard. I see what you're saying about dead code, and I agree that it's better to make cleaner contributions to the code that isn't effectively deprecated. In this case, my last question above I don't understand the answer, which is why I had requested the rationale etc. above.

In order to make this more clear, let me ask the question on the relevant code rather than raising it in the reused function here.

joestringer replied:
OK, looking at the later code I realize my confusion was that I thought the callers were not consistently passing the `ipsec` parameter, but they are actually consistent.

joestringer replied:
Summary: Sure, you don't have to split the commits, but be aware that starting with not splitting the commits meant that I delayed review a lot more because I was more confused about what's going on. At this point it's sunk cost :ship:

pchaigno replied:
Blocking on this. I'm sorry because I understand you want to move on with these changes, but we really need to understand the impact for IPsec coverage here and that should be explained in the commit history. Deprecating ginkgo (where is that documented?) cannot be an excuse to have a poor commit history.

joestringer reviewed (commented):
Thanks for picking this up! I noticed just some minor nits around tidying up some of the references to kernel versions, see the comments below.

pchaigno commented on test/helpers/kubectl.go:
```diff
@@ -2474,7 +2474,7 @@ func (kub *Kubectl) overwriteHelmOptions(options map[string]string) error {
-		if RunsOn419OrLaterKernel() {
+		if RunsOnNetNextKernel() {
 			opts["bpf.masquerade"] = "true"
 			opts["enableIPv6Masquerade"] = "false"
```
The commit description describes the change for 4.19 (now testing KPR=disabled). But this is also changing the behavior for 5.4, no? Do we want to stop testing those features on 5.4? The focus of 5.4 is complexity coverage, so I would expect us to still want to cover BPF masquerading there.

brb replied:
Good catch. I did the code change when I thought that the 5.4 runs only the verifier suite.

pchaigno commented on test/helpers/kubectl.go:
```diff
@@ -2484,19 +2484,17 @@ func (kub *Kubectl) overwriteHelmOptions(options map[string]string) error {
-	if RunsOn419OrLaterKernel() {
+	if RunsOnNetNextKernel() {
 		// To enable SA for both cases when KPR is enabled and disabled
 		addIfNotOverwritten(options, "sessionAffinity", "true")
```
Ditto regarding 5.4.

pchaigno commented on test/helpers/utils.go:
```diff
@@ -578,7 +578,7 @@ func DoesNotRunOnEKS() bool {
 // kube-proxy replacement. Note that kube-proxy may still be running
 // alongside Cilium.
 func RunsWithKubeProxyReplacement() bool {
-	return RunsOnGKE() || RunsOn419OrLaterKernel()
```
Ditto. I think we still want to run KPR on 5.4.

pchaigno commented on test/k8s/services.go:
```diff
@@ -196,8 +196,11 @@ var _ = SkipDescribeIf(helpers.RunsOn54Kernel, "K8sDatapathServicesTest", func()
 					"encryption.enabled": "true",
+					// Until https://github.com/cilium/cilium/issues/23461
+					// has been fixed, we need to disable IPv6 masq
+					"enableIPv6Masquerade": "false",
```
Please have a separate commit with rationale for this as well.

pchaigno reviewed (changes requested):
The commit description doesn't currently answer the "Why?" for several of the changes in this PR.

brb committed:
test: Get rid of 4.9 pipeline

In v1.13 we stopped supporting the 4.9 kernel [1]. Therefore, we can
remove the 4.9 CI pipeline.

Previously, the 4.9 CI used to run with kube-proxy and KPR=disabled,
while the 4.19 CI with kube-proxy and KPR=strict (aka mixed mode). This
commit sets KPR=disabled in the 4.19 CI. The mixed mode is tested in
the new ci-datapath GHA (minus the IPv6 until [2] has been resolved;
this is fine).

[1]: https://github.com/cilium/cilium/issues/22116
[2]: https://github.com/cilium/cilium-cli/issues/1139

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
test: Disable IPv6 masq for IPsec

[1] has all the context. The breakage rose to the surface because in the
previous commit we set KPR=disabled in the 4.19, and iptables started
doing the loadbalancing.

[1]: https://github.com/cilium/cilium/issues/23461

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
test: Do not test E/W eTP=Local with IPsec

[1] has all the context. The breakage rose to the surface because in the
previous commit we set KPR=disabled in the 4.19, and iptables started
doing the loadbalancing.

[1]: https://github.com/cilium/cilium/issues/23481

Signed-off-by: Martynas Pumputis <m@lambda.lt>
